src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java

```diff
@@ -0,0 +1,457 @@
+                        })
+                .isPresent();
+    }
+
+    /**
+     * Checks the identifier ast.
+     *
+     * @param identAst DetailAst of type {@link TokenTypes#IDENT}
+     */
+    private void checkIdentifierAst(DetailAST identAst) {
+        for (VariableDesc variableDesc : variables) {
+            if (identAst.getText().equals(variableDesc.getName())
+                    && !isLeftHandSideValue(identAst)) {
+                variableDesc.registerAsUsed();
+                break;
+            }
+        }
+    }
+
+    /**
+     * Find the scope of variable.
+     *
+     * @param variableDef DetailAst of type {@link TokenTypes#VARIABLE_DEF}
+     * @return scope of variableDef
+     */
+    private static DetailAST findScopeOfVariable(DetailAST variableDef) {
+        final DetailAST result;
+        final DetailAST parentAst = variableDef.getParent();
+        if (parentAst.getType() == TokenTypes.SLIST) {
+            result = parentAst;
+        }
+        else {
+            final DetailAST grandParent = parentAst.getParent();
+            final DetailAST scopeAst = grandParent.findFirstToken(TokenTypes.SLIST);
+            if (scopeAst == null) {
+                result = grandParent;
+            }
+            else {
+                result = scopeAst;
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Checks whether the DetailAst of type {@link TokenTypes#IDENT} is
+     * used as left-hand side value. An identifier is being used as a left-hand side
+     * value if it is used as the left operand of an assignment or as an
+     * operand of a stand-alone increment or decrement.
+     *
+     * @param identAst DetailAst of type {@link TokenTypes#IDENT}
+     * @return True if identAst is used as a left-hand side value otherwise false.
+     */
+    private static boolean isLeftHandSideValue(DetailAST identAst) {
+        final DetailAST parent = identAst.getParent();
+        return isStandAloneIncrementOrDecrement(identAst)
+                || parent.getType() == TokenTypes.ASSIGN
+                && identAst != parent.getLastChild();
+    }
+
+    /**
+     * Checks whether the DetailAst of type {@link TokenTypes#IDENT} is used as
+     * an operand of a stand-alone increment or decrement.
+     *
+     * @param identAst DetailAst of type {@link TokenTypes#IDENT}
+     * @return True if identAst is used as an operand of stand-alone
+     *         increment or decrement otherwise false
+     */
+    private static boolean isStandAloneIncrementOrDecrement(DetailAST identAst) {
+        final DetailAST parent = identAst.getParent();
+        final DetailAST grandParent = parent.getParent();
+        return TokenUtil.isOfType(parent, INCREMENT_AND_DECREMENT_TOKENS)
+                && TokenUtil.isOfType(grandParent, TokenTypes.EXPR)
+                && (!TokenUtil.isOfType(grandParent.getParent(),
+                TokenTypes.ELIST, TokenTypes.INDEX_OP, TokenTypes.ASSIGN)
```

Comment:
Could we extract this as some boolean with descriptive name?

Reply:
Can be but the condition is quite complex for 30-40 characters, this line basically means-

- Is incremented or decremented and used as an argument (`foo.bar(++i);`)
- Is incremented or decremented and used for accessing array (`arr[++i]`)
- Is incremented or decremented and assigned to another variable (`int a = ++i;`)

`isUsedAsArgOrArrAccessorOrRHSAssignment(..)`
Doesn't look or sound good to me but I am open to suggestions

Reply:
All are children of `EXPR`, what do you think about `isUsedInExpression` ?

Reply:
`isUsedInExpr` would be misguiding, consider this example-
```
int a = 12;
++a;
```
parses as-
```
|--EXPR -> EXPR [9:8]
    `--INC -> ++ [9:8]
        `--IDENT -> a [9:10]
```
So `isUsedInExpression(..)` won't work but I get it debugging becomes quite tough with a blob of conditions, what about this-
```
private static boolean isIncrementedOrDecrementedOnRightHandSide(DetailAST grandParent) {
        return TokenUtil.isOfType(grandParent.getParent(),
                TokenTypes.ELIST, TokenTypes.INDEX_OP, TokenTypes.ASSIGN)
                && grandParent.getParent().getParent().getType() != TokenTypes.FOR_ITERATOR;
}
```
as all these conditions imply that the var is being incremented on the right-hand side of the expression. We can furthermore specify in the Javadoc comment what this actually means. Also, we can MAYBE replace IncrementedOrDecremented by variated or something [alike](https://english.stackexchange.com/questions/487090/increment-decrement-is-there-a-neutral-noun-like-crement)

Reply:
> would be misguiding, consider this example...

According to logic in your check, the variable in your example is not "used". 

> Also, we can MAYBE replace IncrementedOrDecremented by variated or something alike...

A general term would just be postfix and prefix, at least this is shorter. `isPrefixOrPostfixOnRightHandSide`?

Either solution is good,

Reply:
extracted to `isPrefixOrPostfixOnLeftHandSide` as I would have to negate the right-hand side condition (`!isPrefixOrPostfixOnRightHandSide(..)`).

Reply:
> I would have to negate the right-hand side condition

Please explain why this is problematic. This makes sense to me, as you've described above.

return value of `isStandAloneIncrementOrDecrement` idea is the following, right?  

"variable should have parent that is inc/dec token AND variable should be part of an expression AND variable should not be used in that expression"

Reply:
> "variable should have parent that is inc/dec token AND variable should be part of an expression AND variable should not be used in that expression"

If this is true, then why not:

```java
        return TokenUtil.isOfType(parent, INCREMENT_AND_DECREMENT_TOKENS)
                && TokenUtil.isOfType(grandParent, TokenTypes.EXPR)
                && !isUsedInExpression(identAst);
```
?

Reply:
> Please explain why this is problematic

It's fine, changed it to `isPrefixOrPostfixOnRightHandSide(..)`, had a faint memory that some check used to show an error when a method is just used once and is negated in that usage.

> variable should have parent that is inc/dec token AND variable should be part of an expression AND variable should not be used in that expression

This line sounds a bit confusing to me.

>  variable should have parent that is inc/dec token AND variable should be part of an expression 

it is fine till this point but the condition and should not be used in that expression sounds odd to me,
the word "used" would be confusing to understand as in this context it means `isConsideredUsedInExpr(..)` or specify in javadoc what used means here.

P.S.- My logic might not be making any sense but then it's just me

Reply:
```java
        x++; // not in expression, so not "used"
        System.out.println(y); // in method call expression, "used" (ELIST)
        int a = arr[w++]; // in assignment expression, "used" (INDEX_OP)
        int b = ++z; // in assignment expression, "used" (ASSIGN)

```

```
        |       |--EXPR -> EXPR [6:9]
        |       |   `--POST_INC -> ++ [6:9]
        |       |       `--IDENT -> x [6:8]
        |       |--SEMI -> ; [6:11]
        |       |--EXPR -> EXPR [7:26]
        |       |   `--METHOD_CALL -> ( [7:26]
        |       |       |--DOT -> . [7:18]
        |       |       |   |--DOT -> . [7:14]
        |       |       |   |   |--IDENT -> System [7:8]
        |       |       |   |   `--IDENT -> out [7:15]
        |       |       |   `--IDENT -> println [7:19]
        |       |       |--ELIST -> ELIST [7:27]
        |       |       |   `--EXPR -> EXPR [7:27]
        |       |       |       `--IDENT -> y [7:27]
        |       |       `--RPAREN -> ) [7:28]
        |       |--SEMI -> ; [7:29]
        |       |--VARIABLE_DEF -> VARIABLE_DEF [8:8]
        |       |   |--MODIFIERS -> MODIFIERS [8:8]
        |       |   |--TYPE -> TYPE [8:8]
        |       |   |   `--LITERAL_INT -> int [8:8]
        |       |   |--IDENT -> a [8:12]
        |       |   `--ASSIGN -> = [8:14]
        |       |       `--EXPR -> EXPR [8:19]
        |       |           `--INDEX_OP -> [ [8:19]
        |       |               |--IDENT -> arr [8:16]
        |       |               |--EXPR -> EXPR [8:21]
        |       |               |   `--POST_INC -> ++ [8:21]
        |       |               |       `--IDENT -> w [8:20]
        |       |               `--RBRACK -> ] [8:23]
        |       |--SEMI -> ; [8:24]
        |       |--VARIABLE_DEF -> VARIABLE_DEF [9:8]
        |       |   |--MODIFIERS -> MODIFIERS [9:8]
        |       |   |--TYPE -> TYPE [9:8]
        |       |   |   `--LITERAL_INT -> int [9:8]
        |       |   |--IDENT -> b [9:12]
        |       |   `--ASSIGN -> = [9:14]
        |       |       `--EXPR -> EXPR [9:16]
        |       |           `--INC -> ++ [9:16]
        |       |               `--IDENT -> z [9:18]

```

Reply:
>should not be used in that expression sounds odd to me

Yes, I understand. It is "used" in higher ancestor expression, not grandparent

Reply:
> // not in expression, so not "used"

This line is difficult to comprehend with respect to this AST

```
        |       |--EXPR -> EXPR [6:9]
        |       |   `--POST_INC -> ++ [6:9]
        |       |       `--IDENT -> x [6:8]
```
Someone who wants to make changes to this code in the future will not understand 
`not in expression, so not "used"` as it's right there "inside" the expression, maybe make it more clear by adding `not in the right-hand side expression, so not "used"` where we can specify what is considered to be a right-hand side expression.

Reply:
I edited above comment, take a look

Reply:
What about `hasExpressionAncestor`? I am good with current name, though, just trying to think of options.

Reply:
> It is "used" in higher ancestor expression, not grandparent.

Ther is no expr node as an ancestor for `int b = ++z;`
```
        |--VARIABLE_DEF -> VARIABLE_DEF [12:8]
        |       |   |--MODIFIERS -> MODIFIERS [12:8]
        |       |   |--TYPE -> TYPE [12:8]
        |       |   |   `--LITERAL_INT -> int [12:8]
        |       |   |--IDENT -> b [12:12]
        |       |   `--ASSIGN -> = [12:14]
        |       |       `--EXPR -> EXPR [12:16]
        |       |           `--INC -> ++ [12:16]
        |       |               `--IDENT -> z [12:18]

```

Reply:
`=` is expression by java grammar, but not checkstyle grammar :) This could be confusing difference for future maintainers, I agree

Reply:
`isPrefixOrPostfixUsed` sounds good to me

Reply:
I agree.

Reply:
Resolved.

Reply:
Resolved.

Reply:
Changed it to `isPrefixOrPostfixVariableUsed` as it could also mean `isPrefixOrPostfixOperatorUsed`

Reply:
Just had another thought; what about `isStandAloneUsed`? This avoids having both increment/decrement and prefix/postfix nomenclature in the same method, and continues to use the same term, 'standAlone'?

Reply:
`isStandAloneUsed` sounds like it's talking about whether an operator or property called `StandAlone` is used, I think the current one is more expressive.

Reply:
That makes sense. It would probably be good to use `isIncrementOrDecrementVariableUsed` to be consistent, it is only a few letters difference now.

Either way, you can resolve this.

Reply:
Done

Reply:
Resolved.